### DISCUSSION TITLE
define kafkaconnect worker consumer configurations

### DISCUSSION THREAD

**Discussion by cstmgl** [strimzi/strimzi-kafka-operator#5020] [Q&A] [answered]
Hi All,
I have a question about how to define the Kafka Connect worker consumer configurations.
I've been reading a little and I think they need to be defined at the worker level and for example I could override them at each  Connector level if I would define the override policy also at the worker level
`connector.client.config.override.policy`

Now my assumption is that the worker level would mean that I defined this together with my KafkaConnect CRD but I'm not sure this works because I get some warnings in the logs that some of this properties are defined but do not exist (when I look at the logs in the instance).

So I just wanted to know if there is any example of this, I also did a quick browse on the documentation but did not find something specific to this, but I might have overlooked something.

Thank you for any feedback

just for reference this are the warnings I mean I get on startup of my pods:
![image](https://user-images.githubusercontent.com/33542778/119453321-98d6d880-bd37-11eb-8d56-1d437494e971.png)

**Comment by scholzj** [comment, marked as answer]
I think you can just ignore these warnings. Kafka just takes the configuration and passes it to every kind of component. And not all of the options are valid for every components. So they produce these kind of warnings. But the component which supports them will find them and use them.

**Reply by cstmgl** [reply]
thank you for the feedback, I'm giving it a try and will try to see if this works as I expect it to work

**Reply by tusharbhasme** [reply]
@scholzj I need to do smthing similar. I have to deploy many connectors, and need to use a different ACL(KafkUser) per connector as each connector will have its own set of topics. For this, I assume I need to specify:
``` kafka-connect.yaml
kind: KafkaConnect
spec:
  config:
    connector.client.config.override.policy=All
```
``` sink-connector.yaml
kind: KafkaConnector
spec:
  config:
    consumer.override.ssl.keystore.location=/var/private/ssl/sink-user.keystore.jks
    consumer.override.ssl.keystore.password=connector1234
    consumer.override.ssl.key.password=connector1234
```
My question is, how should I pass the user keystore info stored in the sink-user's secret to KafkaConnector? I don't see any way to pass the secret info in KafkaConnector schema.

**Reply by scholzj** [reply]
You can use this to mount additional secrets: https://strimzi.io/docs/operators/latest/full/using.html#type-ExternalConfiguration-reference ... and from 0.24 we will in addition support https://github.com/strimzi/kafka-kubernetes-config-provider

**Reply by tusharbhasme** [reply]
We are currently on 0.21.1, and are getting this ClassNotFoundException for DirectoryConfigProvider:
```
2021-06-11 18:05:35,706 ERROR ClassNotFoundException exception occurred: org.apache.kafka.common.config.provider.DirectoryConfigProvider (org.apache.kafka.connect.ru │
│ 2021-06-11 18:05:35,713 ERROR Stopping due to error (org.apache.kafka.connect.cli.ConnectDistributed) [main]                                                          │
│ org.apache.kafka.common.config.ConfigException: Invalid value java.lang.ClassNotFoundException: org.apache.kafka.common.config.provider.DirectoryConfigProvider for c │
│     at org.apache.kafka.common.config.AbstractConfig.instantiateConfigProviders(AbstractConfig.java:545)                                                              │
│     at org.apache.kafka.common.config.AbstractConfig.resolveConfigVariables(AbstractConfig.java:484)                                                                  │
│     at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:107)                                                                                  │
│     at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:129)                                                                                  │
│     at org.apache.kafka.connect.runtime.WorkerConfig.<init>(WorkerConfig.java:452)                                                                                    │
│     at org.apache.kafka.connect.runtime.distributed.DistributedConfig.<init>(DistributedConfig.java:393)                                                              │
│     at org.apache.kafka.connect.cli.ConnectDistributed.startConnect(ConnectDistributed.java:93)                                                                       │
│     at org.apache.kafka.connect.cli.ConnectDistributed.main(ConnectDistributed.java:78)  
```
We also tried FileConfigProvider, but it is not able to read secrets using that.
```
org.apache.kafka.common.config.ConfigException: Could not read properties from file /opt/kafka/external-configuration/sink-connector-config                           │
│     at org.apache.kafka.common.config.provider.FileConfigProvider.get(FileConfigProvider.java:92)                                                                     │
│     at org.apache.kafka.common.config.ConfigTransformer.transform(ConfigTransformer.java:103)                                                                         │
│     at org.apache.kafka.connect.runtime.WorkerConfigTransformer.transform(WorkerConfigTransformer.java:58)                                                            │
│     at org.apache.kafka.connect.runtime.WorkerConfigTransformer.transform(WorkerConfigTransformer.java:53)                                                            │
│     at org.apache.kafka.connect.runtime.AbstractHerder.validateConnectorConfig(AbstractHerder.java:336)                                                               │
│     at org.apache.kafka.connect.runtime.AbstractHerder.lambda$validateConnectorConfig$1(AbstractHerder.java:326)                                                      │
│     at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)                                                                              │
│     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)                                                                                             │
│     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)                                                                      │
│     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)                                                                      │
│     at java.base/java.lang.Thread.run(Thread.java:834) 
```

**Reply by scholzj** [reply]
I think DirectoryConfigProvider was maybe added in Kafka 2.7.0 or something like that? The FileConfigProvider should work fine, that exists for a long time.

**Reply by tusharbhasme** [reply]
It seems the secret needs to be in properties format. Currently, the secret is being mounted as a file for each entry, while the FileConfigProvider expects a single properties file with entry for each of that in the secret:
```
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
stringData:
  connector.properties: |-
    userKey: <key>
    userPassword: <password>
```
 Is there a way to change the secret's format as above when we create KafkaUser?

**Reply by scholzj** [reply]
Secret cannot be mounted as a properties file. You simple either need to upgrade and use the DirectoryConfigProvider or create your secret with a properties inside so that it works with the FileConfigProvider.

**Reply by tusharbhasme** [reply]
I was able to get it working. Had to make a little change in the connector definition though. The [documentation](https://strimzi.io/docs/operators/latest/full/using.html#type-ExternalConfiguration-reference) specified:
```
consumer.override.ssl.keystore.location: "${directory:/opt/kafka/external-configuration/sink-connector-config:user.p12}"
```
ie the content of the secret.
I had to actually provide just the mounted location of the secret to make it work:
```
consumer.override.ssl.keystore.location: "/opt/kafka/external-configuration/sink-connector-config/user.p12"
```
Specifying the first way didn't work and kept throwing errors. How would it work with KubernetesSecretConfigProvider in 0.24? I am assuming it would again provide the actual content of the key, which did not work out for me.

**Reply by scholzj** [reply]
Is `{directory:/opt/kafka/external-configuration/sink-connector-config:user.p12}` in the documentation somewhere? That would work only with PEM certificates.

**Reply by tusharbhasme** [reply]
No, the documentation uses `user.key`, however, in my case, I need to use a separate `user` per connector, and hence need to use the property `consumer.override.ssl.keystore.location` for which  I am not able to specify the PEM key. It keeps failing for `org.apache.kafka.common.errors.InvalidConfigurationException: SSL key store password cannot be specified with PEM format, only key password may be specified` even though I did not specify it. It seems it is getting it from somewhere else on top.

**Reply by scholzj** [reply]
Well, the PEM key does not have password. So you just need to fix the configuration. Check the Kafka docs or the Strimzi Client examples which uses the PEM keys. Anyway, you can use whatever you want - the p12 file is there to be used as well. So no problem with that.

The KubernetesSecretConfigProvider will provide the content of the filed, not path.

**Reply by tusharbhasme** [reply]
Yeah, and I really do want to use the content instead of the file so that I can use `KubernetesSecretConfigProvider` when available. I can't find out where it is getting the keystore password from. client examples are also setting the same properties I have set. As per kafka's source code, it has to be the worker's properties, but as per this configuration, it is not using any password either:
```yaml
  authentication:
    type: tls
    certificateAndKey:
      certificate: user.crt
      key: user.key
      secretName: test-user
```
Even if it is setting the password behind the scenes, how can we make it not do so?

**Reply by scholzj** [reply]
I guess it is override ... so it probably inherits the original options and you are just overriding them?

**Reply by tusharbhasme** [reply]
Right. I am able to override the `keystore.location`, and can override `keystore.password`, but not remove it if I need to use PEM format.

**Reply by scholzj** [reply]
I do not remember the Kafka code ... have you tried to set it to empty string?

**Reply by tusharbhasme** [reply]
Yep. didn't work.

**Reply by tusharbhasme** [reply]
Does KafkaConnect set the keystore password while reading the TLS secrets, though we are not specifying the password anywhere in the configuration?

**Reply by scholzj** [reply]
You specify the TLS certificate which should be used in the keystore in the `authentication` section. Strimzi generates the keystore out of it and sets the values.

**Reply by tusharbhasme** [reply]
Oh, seems it must also be setting some password that is getting passed on to the connector. Would it be a Kafka bug where Connector is not able to override using different PEM certs when Connect is setup for TLS with keystore?

**Reply by tusharbhasme** [reply]
Or could Strimzi provide an option to specify Keystore or Key to be used in TLS, instead of generating the Keystore behind the scenes?

**Reply by scholzj** [reply]
Probably not ... I guess it expects you to override options and not remove them. Using PKCS12 on one level and PEM on the other is a bit weird use-case. @tombentley WDYT?

I expect that we in the future move to using the PEM certificates as well, which might allow this. But definitely not in Strimzi 0.24. Until then I guess you need to use the PKCS12 file mounted using the `externalConfigiration` in `KafkaConnect`.

Another option - which I know many users prefer - is to actually run one Connect per connector. In whihc case you do not care about the overrides at all.

**Reply by scholzj** [reply]
> Or could Strimzi provide an option to specify Keystore or Key to be used in TLS, instead of generating the Keystore behind the scenes?

No, there is definitely no plan for this. The costs for maintaining it would be too big given what it really brings and the number of workarounds.

**Reply by tusharbhasme** [reply]
Should I raise a feature request to use PEM certs while setting up TLS for Connect workers?

**Reply by scholzj** [reply]
You can if you want. But it is not needed, we have it in our plans anyway.

**Reply by tusharbhasme** [reply]
BTW, what is the planned release date of Strimzi 0.24?

**Reply by scholzj** [reply]
RC1 was released yesterday. So probably some time next week, depending on how many issues we find in the RC1. You can help by testing it: https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.24.0-rc1 ;-)

**Reply by tusharbhasme** [reply]
Sure, how do I help?

**Reply by scholzj** [reply]
You can just give it a try and see if it works for you, if you run into any issues etc.

**Reply by tusharbhasme** [reply]
Sadly we are still on 0.21 without a plan to upgrade yet. :-(
In the meantime, can you share the issue, that we can track, under which the Connect's TLS setup will be changed to use PEM format?

**Reply by scholzj** [reply]
I do not think it has an issue, that is why I said you can open one if you want - sorry if I was not clear. But it is part of a bigger plan to get rid of the certificate processing in the images (and ideally get rid of OpenSSL).

**Reply by tusharbhasme** [reply]
https://github.com/strimzi/strimzi-kafka-operator/issues/5189
